### PR TITLE
Detect subtype in urls

### PR DESCRIPTION
**Pull request by Ark2307**

### CODE REVIEW

File: apps/api-runtime/src/main/java/com/akto/runtime/APICatalogSync.java [hunk 1]
```diff
@@ -574,18 +575,19 @@ public static boolean areBothUuidUrls(URLStatic newUrl, URLStatic deltaUrl, URLT
...
         SuperType[] b = mergedTemplate.getTypes();
         for (int idx =0 ; idx < b.length; idx++) {
             SuperType c = b[idx];
             if (Objects.equals(c, SuperType.STRING) && o.length > idx) {
                 String val = n[idx];
-                if(!isAlphanumericString(val) || !isAlphanumericString(o[idx])) {
+                SubType newSubType = getTokenSubType(n[idx]);
+                SubType oldSubType = getTokenSubType(o[idx]);
+                boolean isAlphaNumeric = isAlphanumericString(val) && isAlphanumericString(o[idx]);
+                boolean isUserRegex = isValidSubtype(oldSubType) && isValidSubtype(newSubType);
```
Comment (ankush-jain-akto) [inline apps/api-runtime/src/main/java/com/akto/runtime/APICatalogSync.java:588]: Both should be equal and non-null

File: apps/api-runtime/src/main/java/com/akto/runtime/APICatalogSync.java [hunk 2]
```diff
@@ -574,18 +575,19 @@ public static boolean areBothUuidUrls(URLStatic newUrl, URLStatic deltaUrl, URLT
...
-
         String[] n = tokenize(newUrl.getUrl());
         String[] o = tokenize(deltaUrl.getUrl());
         SuperType[] b = mergedTemplate.getTypes();
         for (int idx =0 ; idx < b.length; idx++) {
             SuperType c = b[idx];
             if (Objects.equals(c, SuperType.STRING) && o.length > idx) {
                 String val = n[idx];
-                if(!isAlphanumericString(val) || !isAlphanumericString(o[idx])) {
+                SubType newSubType = getTokenSubType(n[idx]);
```
Comment (ankush-jain-akto) [inline apps/api-runtime/src/main/java/com/akto/runtime/APICatalogSync.java:585]: If `newSubType` comes out to be null, continue.

File: apps/api-runtime/src/main/java/com/akto/runtime/APICatalogSync.java [hunk 3]
```diff
@@ -613,6 +615,24 @@ public static boolean isAlphanumericString(String s) {
...
+        for (CustomDataType customDataType: SingleTypeInfo.getCustomDataTypesSortedBySensitivity(Context.accountId.get())) {
+            if (!customDataType.isActive()) continue;
+            boolean result = false;
+            Conditions valueConditions =  customDataType.getValueConditions();
+            
+            if(valueConditions !=null){
+                result = valueConditions.validate(token);
+            }
+            if (result) return customDataType.toSubType();
+        }
```
Comment (ankush-jain-akto) [inline apps/api-runtime/src/main/java/com/akto/runtime/APICatalogSync.java:628]: We don't need this logic here. Please use `findSubType` directly

File: apps/api-runtime/src/main/java/com/akto/runtime/APICatalogSync.java [hunk 4]
```diff
@@ -613,6 +615,24 @@ public static boolean isAlphanumericString(String s) {
...
+        return KeyTypes.findSubType(token, "", null,true);
+    }
+
+    private static Boolean isValidSubtype(SubType subType){
```
Comment (ankush-jain-akto) [inline apps/api-runtime/src/main/java/com/akto/runtime/APICatalogSync.java:632]: small 'b' in Boolean

File: apps/api-runtime/src/main/java/com/akto/runtime/APICatalogSync.java [hunk 5]
```diff
@@ -633,6 +653,9 @@ public static URLTemplate tryMergeUrls(URLStatic dbUrl, URLStatic newUrl) {
             String tempToken = newTokens[i];
             String dbToken = dbTokens[i];
 
+            SubType dbSubType = getTokenSubType(dbTokens[i]);
```
Comment (ankush-jain-akto) [inline apps/api-runtime/src/main/java/com/akto/runtime/APICatalogSync.java:656]: This same check isn't required again.

File: apps/api-runtime/src/main/java/com/akto/runtime/APICatalogSync.java (hunk 3 above)
Comment (Ark2307) [reply apps/api-runtime/src/main/java/com/akto/runtime/APICatalogSync.java:628]: done

File: apps/api-runtime/src/main/java/com/akto/runtime/APICatalogSync.java (hunk 1 above)
Comment (Ark2307) [reply apps/api-runtime/src/main/java/com/akto/runtime/APICatalogSync.java:588]: done

File: libs/dao/src/main/java/com/akto/dto/type/KeyTypes.java [hunk 6]
```diff
@@ -198,6 +186,30 @@ public static SubType findSubType(Object o,String key, ParamId paramId) {
         return SingleTypeInfo.OTHER;
     }
 
+    public static SubType findSubType(Object o,String key, ParamId paramId, Boolean executeCheckForSubtypes){
```
[the remark this answers is not in the document]
Comment (Ark2307) [reply libs/dao/src/main/java/com/akto/dto/type/KeyTypes.java]: done

File: apps/api-runtime/src/main/java/com/akto/runtime/APICatalogSync.java (hunk 4 above)
Comment (Ark2307) [reply apps/api-runtime/src/main/java/com/akto/runtime/APICatalogSync.java:632]: done

**Pull request closed without merging**